test_vm/src/util.rs

```diff
@@ -177,54 +179,105 @@ pub fn precommit_sectors(
+            );
+            let expect = ExpectInvocation {
+                to: mid,
+                method: MinerMethod::PreCommitSectorBatch as u64,
+                params: Some(
+                    serialize(
+                        &PreCommitSectorBatchParams { sectors: param_sectors },
+                        "precommit batch params",
+                    )
+                    .unwrap(),
+                ),
+                subinvocs: Some(invocs),
                 ..Default::default()
-            });
-            sector_idx += 1;
-            j += 1;
-        }
-        if param_sectors.len() > 1 {
-            invocs.push(invoc_net_fee(aggregate_pre_commit_network_fee(
-                param_sectors.len() as i64,
-                &TokenAmount::zero(),
-            )));
-        }
-        if expect_cron_enroll && msg_sector_idx_base == 0 {
-            invocs.push(invoc_first());
+            };
+            expect.matches(v.take_invocations().last().unwrap())
+        } else {
+            let mut param_sectors = Vec::<SectorPreCommitInfo>::new();
+            let mut j = 0;
+            while j < batch_size && sector_idx < count {
+                let sector_number = sector_number_base + sector_idx;
+                param_sectors.push(SectorPreCommitInfo {
+                    seal_proof,
+                    sector_number,
+                    sealed_cid: make_sealed_cid(format!("sn: {}", sector_number).as_bytes()),
+                    seal_rand_epoch: v.get_epoch() - 1,
+                    deal_ids: vec![],
+                    expiration,
+                    unsealed_cid: CompactCommD::new(None),
+                });
+                sector_idx += 1;
+                j += 1;
+            }
+            if param_sectors.len() > 1 {
+                invocs.push(invoc_net_fee(aggregate_pre_commit_network_fee(
+                    param_sectors.len() as i64,
+                    &TokenAmount::zero(),
+                )));
+            }
+            if expect_cron_enroll && msg_sector_idx_base == 0 {
+                invocs.push(invoc_first());
+            }
+
+            apply_ok(
+                v,
+                worker,
+                maddr,
+                TokenAmount::zero(),
+                MinerMethod::PreCommitSectorBatch2 as u64,
+                PreCommitSectorBatchParams2 { sectors: param_sectors.clone() },
+            );
+            let expect = ExpectInvocation {
+                to: mid,
+                method: MinerMethod::PreCommitSectorBatch2 as u64,
+                params: Some(
+                    serialize(
+                        &PreCommitSectorBatchParams2 { sectors: param_sectors },
+                        "precommit batch params",
+                    )
+                    .unwrap(),
+                ),
+                subinvocs: Some(invocs),
+                ..Default::default()
+            };
+            expect.matches(v.take_invocations().last().unwrap())
```

Comment:
Nit: I wish we could extract more re-usable code here, but I can see the types are going to make it difficult.

Reply:
Yeah, types make it difficult. I would have to go with a pattern I used in the unit tests harness but it isn't much cleaner.